overwatch/src/services/runner.rs

```diff
@@ -0,0 +1,350 @@
+
+        ServiceRunnerHandle {
+            service_handle,
+            runner_join_handle,
+        }
+    }
+
+    async fn run_<Service>(self)
+    where
+        Service: ServiceCore<RuntimeServiceId, Settings = Settings, State = State, Message = Message>
+            + 'static,
+        StateOp: Clone,
+    {
+        let Self {
+            mut service_resources,
+            mut service_lifecycle_phase,
+        } = self;
+
+        // Handles to hold the Service and StateHandle tasks
+        let mut service_task_handle: Option<_> = None;
+        let mut state_handle_task_handle: Option<_> = None;
+
+        while let Some(lifecycle_message) = service_resources.lifecycle_handle.next().await {
+            match lifecycle_message {
+                LifecycleMessage::Start(finished_signal_sender) => {
+                    if service_lifecycle_phase == LifecyclePhase::Started {
+                        info!("Service is already running.");
+                    } else {
+                        Self::handle_start::<Service>(
+                            &mut service_resources,
+                            &mut service_task_handle,
+                            &mut state_handle_task_handle,
+                        );
+                        service_lifecycle_phase = LifecyclePhase::Started;
+                    }
+
+                    // TODO: Sending a different signal could be handy to differentiate whether
+                    //  the service was already started or not.
+                    if let Err(error) = finished_signal_sender.send(()) {
+                        dbg!(
+                            "Error while sending the LifecycleMessage::Start signal: {}.",
+                            error
+                        );
+                    }
+                }
+                LifecycleMessage::Stop(finished_signal_sender) => {
+                    if service_lifecycle_phase == LifecyclePhase::Stopped {
+                        info!("Service is already stopped.");
+                    } else {
+                        Self::handle_stop(
+                            &mut service_task_handle,
+                            &mut state_handle_task_handle,
+                            &mut service_resources,
+                        )
+                        .await;
+                        service_lifecycle_phase = LifecyclePhase::Stopped;
+                    }
+
+                    // TODO: Sending a different signal could be handy to differentiate whether
+                    //  the service was already stopped or not.
+                    if let Err(error) = finished_signal_sender.send(()) {
+                        dbg!("Error while sending the LifecycleMessage::Stop finished signal: {}. Likely due to the receiver being already dropped in the Service::run task.", error);
+                    }
+                }
+            }
+        }
+    }
+
+    fn handle_start<Service>(
```

Comment:
Can't we put any of the stuff we produce inside here as part of the `LifecyclePase::Started` variant?

Reply:
What do you mean?

Reply:
Are the variables that are initialized as part of the `handle_start` function needed also when the service is not starting? I am suggesting to group them as part of the lifecycle phase enum, so that they are only accessible when the service is running. If that's not the case, then please disregrad this comment.

Reply:
I think we might be able to move `service_resources_handle` and `inbound_relay`, but not `initial_state`.
That being said, I kinda lean more towards making two different APIs for `ServiceResources`. One for when the `Service` is stopped and another when for when it's running.

Reply:
Yes, a different API might also make sense, but still I think it would not hurt to rely on the compiler for stuff that is meant to exist only when the service is in a specific state, but only if easily achievable.

Reply:
Yeah I checked the code again and it probably won't add too much value, so let's keep it as it is. Sorry for raising a false positive.

Reply:
No worries, better a quick double check than having to come back to this later :)